projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/listeners/ParameterAttributesGroup.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2017, Oracle and/or its affiliates.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are
+ * permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list of
+ * conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of
+ * conditions and the following disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to
+ * endorse or promote products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.parser.listeners;
+
+import java.util.List;
+
+import com.oracle.truffle.llvm.parser.model.attributes.AttributesGroup;
+import com.oracle.truffle.llvm.parser.model.attributes.KnownAttribute;
+import com.oracle.truffle.llvm.parser.model.attributes.KnownValueAttribute;
+import com.oracle.truffle.llvm.parser.model.attributes.StringAttribute;
+import com.oracle.truffle.llvm.parser.model.attributes.StringValueAttribute;
+import com.oracle.truffle.llvm.parser.records.ParameterAttributeGroupRecord;
+import com.oracle.truffle.llvm.runtime.LLVMLogger;
+
+public final class ParameterAttributesGroup implements ParserListener {
+
+    private final List<AttributesGroup> attributes;
+
+    public ParameterAttributesGroup(List<AttributesGroup> attributes) {
+        this.attributes = attributes;
+    }
+
+    @Override
+    public void record(long id, long[] args) {
+        switch ((int) id) {
+            case 3:
+                AttributesGroup group = new AttributesGroup(args[0], args[1]);
+                attributes.add(group);
+
+                int i = 2;
+                while (i < args.length) {
+                    long type = args[i++];
+                    switch ((int) type) {
+                        case 0: // well-known attribute
```

Comment:
Could we have named constants instead of integer literals? Alternatively, to improve code consistency with the rest of the parser, could you create an enum with a 'decode(long)' method?